topology/probes/vpp/vpp.go

```diff
@@ -103,268 +124,508 @@ func (p *Probe) getInterfaceVrfID(ch api.Channel, index uint32) int64 {
+
+	return command, nil
+}
+
+func (p *Probe) OnEdgeAdded(e *graph.Edge) {
+	if edgeType, _ := e.GetFieldString("RelationType"); edgeType != topology.OwnershipLink {
+		return
+	}
+
+	intfNode := p.Ctx.Graph.GetNode(e.Child)
+	if s, _ := intfNode.GetFieldString("VPP.SocketFilename"); s != "" {
+		p.indexSocketFilename(s, intfNode)
+	} else if pid, err := intfNode.GetFieldInt64("InitProcessPID"); err == nil {
+		parent := p.Ctx.Graph.GetNode(e.Parent)
+		if t, _ := parent.GetFieldString("Type"); t == "netns" {
+			if _, found := p.seeds[pid]; !found {
+				seed, err := p.startSeed(pid, parent)
+				if err != nil {
+					p.Ctx.Logger.Errorf("Failed to start seed for %d: %s", pid, err)
+					return
+				}
+				p.seeds[pid] = seed
+			}
+		}
+	}
+}
+
+func (p *Probe) GetABLinks(node *graph.Node) (edges []*graph.Edge) {
+	socketFilename, _ := node.GetFieldString("VPP.SocketFilename")
+	socketFilename, _ = p.getHostSocketFilename(socketFilename, node)
+	nodes, _ := p.slaveMemifs.Get(socketFilename)
+	for _, slave := range nodes {
+		edges = append(edges, topology.NewLink(p.Ctx.Graph, node, slave, "memif", nil))
+	}
+	return
+}
+
+func (p *Probe) GetBALinks(node *graph.Node) (edges []*graph.Edge) {
+	socketFilename, _ := node.GetFieldString("VPP.SocketFilename")
+	socketFilename, _ = p.getHostSocketFilename(socketFilename, node)
+	nodes, _ := p.masterMemifs.Get(socketFilename)
+	for _, master := range nodes {
+		edges = append(edges, topology.NewLink(p.Ctx.Graph, master, node, "memif", nil))
+	}
+	return
+}
+
+func (p *Probe) handleInterfaceEvent(msg *interfaces.SwInterfaceEvent) {
+	p.Ctx.Graph.Lock()
+	if node := p.getInterface(msg.SwIfIndex); node != nil {
+		name, _ := node.GetFieldString("Name")
+		if msg.Deleted > 0 {
+			p.Ctx.Logger.Debugf("Delete interface %v idx %d", name, msg.SwIfIndex)
+			p.eventDelInterface(node)
+		} else {
+			state := interfaceUpDown(msg.AdminUpDown)
+			p.Ctx.Logger.Debugf("ChangeState interface %s idx %d updown %d", name, msg.SwIfIndex, state)
+			p.Ctx.Graph.AddMetadata(node, "State", state)
+		}
+	}
+	p.Ctx.Graph.Unlock()
+}
+
+func (p *Probe) run(ctx context.Context) {
+	ch, err := p.conn.NewAPIChannel()
 	if err != nil {
 		p.Ctx.Logger.Error("API channel error: ", err)
 		return
 	}
+	ch.SetReplyTimeout(HealthCheckReplyTimeout)
 
-	req := &vpe.ShowVersion{}
-	msg := &vpe.ShowVersionReply{}
-	err = ch.SendRequest(req).ReceiveReply(msg)
+	notifChan := make(chan api.Message, 100)
+	sub, err := ch.SubscribeNotification(notifChan, &interfaces.SwInterfaceEvent{})
 	if err != nil {
 		p.Ctx.Logger.Error(err)
 		return
 	}
-	ch.Close()
 
-	metadata := graph.Metadata{
-		"Name":      "vpp",
-		"Type":      "vpp",
-		"Program":   string(msg.Program),
-		"Version":   string(msg.Version),
-		"BuildDate": string(msg.BuildDate),
-	}
+	p.Ctx.Logger.Debugf("Registering for VPP events")
+	p.interfaceEventsEnableDisable(ch, true)
 
-	p.Ctx.Graph.Lock()
-	defer p.Ctx.Graph.Unlock()
+	pollingTicker := time.NewTicker(VPPPollingTime)
```

Comment:
this could/should be moved after line 485 ?

Reply:
Done